game-core/src/main/java/games/strategy/engine/data/GameParser.java

```diff
@@ -1169,26 +1169,60 @@ private void parseAttachments(final Element root, final Map<String, List<String>
       if (foreach.isEmpty()) {
         parseAttachment(current, variables, Collections.emptyMap());
       } else {
-        final List<String> foreachVariables = Splitter.on(":").splitToList(foreach);
-        if (!variables.keySet().containsAll(foreachVariables)) {
-          throw newGameParseException("Attachment has invalid variables in foreach: " + foreach);
+        final List<String> nestedForeach = Splitter.on("^").splitToList(foreach);
+        if (nestedForeach.size() > 2) {
+          throw newGameParseException("Currently only supports using one '^' for nested foreach: " + foreach);
         }
-        final int length = variables.get(foreachVariables.get(0)).size();
-        for (int i = 0; i < length; i++) {
-          final Map<String, String> foreachMap = new HashMap<>();
-          for (final String foreachVariable : foreachVariables) {
-            final List<String> foreachValue = variables.get(foreachVariable);
-            if (length != foreachValue.size()) {
-              throw newGameParseException("Attachment foreach variables must have same number of elements: " + foreach);
+        final List<String> foreachVariables1 = Splitter.on(":").splitToList(nestedForeach.get(0));
```

Comment:
Wouldn't `nestedForeach.get(0)` throw an exception if the list is empty?
You might want to create a pre-condition for this.

Reply:
I don't believe there is really any reasonable way it could be empty at this point as it checks that the string isn't empty with `foreach.isEmpty()` then splits on '^' with `Splitter.on("^").splitToList(foreach)`. So I think the only way it could be empty is if the original foreach string has only '^' characters ("^", "^^^^", etc). Guess I figured no one would be crazy enough to put that as the foreach string but if you think its worth it then I could just update the nestedForeach size check above to `(nestedForeach.isEmpty() || nestedForeach.size() > 2)`.

Reply:
I was more thinking about the case where there's just an empty string in the first place, but yeah adding that simple check would be nice 👍

Reply:
Alright. Updated.